File: backend/app/optical/kinds/lens/physics.py
"""Lens PhysicsOp (Python mirror of frontend/src/optical/kinds/lens/physics.ts).

Op name: abcd_thin_lens
Kind: lens
Required params: focalLengthMm
"""

from __future__ import annotations

import math

from app.optical.beam_ray import BeamRay, vec3_distance
from app.optical.registry import (
    KindEntry,
    PhysicsOp,
    PhysicsOpContext,
    register_kind,
)
from app.solvers.generalized_abcd import m_thin_lens


def apply_abcd_to_q(A: float, B: float, C: float, D: float, q: complex) -> complex:
    """q_out = (A*q + B) / (C*q + D)."""
    return (A * q + B) / (C * q + D)


def abcd_thin_lens_op(ray_in: BeamRay, ctx: PhysicsOpContext) -> list[BeamRay]:
    # Prefer the transition's explicit 5x5 matrix when supplied — this is
    # how plano-convex lenses encode their direction-aware thick-lens
    # matrices (convex→flat vs flat→convex produce different BFL).
    M = None
    if ctx.transfer_matrix is not None and ctx.transfer_matrix[0] == "matrix5x5":
        flat = ctx.transfer_matrix[1]
        import numpy as np
        # Accept either flat list-of-25 (row-major) or list-of-5-rows.
        if isinstance(flat[0], (list, tuple)):
            M = np.array(flat, dtype=float)
        else:
            M = np.array(flat, dtype=float).reshape(5, 5)
    if M is None:
        focal_mm = ctx.params.get("focalLengthMm") or ctx.params.get("focalMm")
        if not isinstance(focal_mm, (int, float)) or not math.isfinite(focal_mm) or focal_mm == 0:
            raise ValueError(
                f"abcd_thin_lens: needs transition matrix5x5 OR valid focalLengthMm; got {focal_mm!r}"
            )
        M = m_thin_lens(float(focal_mm))
    # x sub-block (rows 0/1, cols 0/1)
    Ax, Bx = M[0, 0], M[0, 1]
    Cx, Dx = M[1, 0], M[1, 1]
    # y sub-block (rows 2/3, cols 2/3)
    Ay, By = M[2, 2], M[2, 3]
    Cy, Dy = M[3, 2], M[3, 3]

    qx_out = apply_abcd_to_q(Ax, Bx, Cx, Dx, ray_in.qx)
    qy_out = apply_abcd_to_q(Ay, By, Cy, Dy, ray_in.qy)

    # Propagate chief ray from face_in to face_out (thin-lens approximation).
    thickness_mm = vec3_distance(
        ctx.face_in.position_mm_body_local,
        ctx.face_out.position_mm_body_local,
    )
    new_origin = ray_in.origin + ray_in.direction * thickness_mm

    return [ray_in.replaced(
        origin=new_origin,
        qx=qx_out,
        qy=qy_out,
        path_length_mm=ray_in.path_length_mm + thickness_mm,
    )]


# Registration (runs on import of this module)
register_kind("lens", KindEntry(
    ops={"abcd_thin_lens": abcd_thin_lens_op},
    needs_aperture=True,
))

File: frontend/src/optical/kinds/lens/physics.ts
/**
 * Lens PhysicsOp — thin-lens ABCD transformation on a BeamRay's q-parameter,
 * with the chief ray geometrically propagated from face_in to face_out
 * along its current direction.
 *
 * Wraps the existing `mThinLens` constructor from `generalizedAbcd.ts`
 * (5×5 matrix form). Only the q-parameter update uses the matrix;
 * Jones, power, and direction pass through unchanged (a thin lens is
 * polarization-preserving and lossless in the ideal model).
 *
 * Op name: `abcd_thin_lens`
 * Kind:    `lens`
 * Required params: `focalLengthMm: number`
 */

import {
  type BeamRay,
  vec3Add,
  vec3Distance,
  vec3Scale,
  vec3Sub,
} from "../../beam-ray";
import { type Complex, cAdd, cDiv, cMul } from "../../fiber/gaussian";
import { mThinLens } from "../../generalizedAbcd";
import {
  type PhysicsOp,
  type PhysicsOpContext,
  registerKind,
} from "../../registry";

// ---------------------------------------------------------------------------
// q-parameter ABCD update
// ---------------------------------------------------------------------------

/** Apply a 2×2 ABCD matrix to a complex q-parameter:
 *  q_out = (A·q + B) / (C·q + D). */
export function applyAbcdToQ(
  A: number, B: number, C: number, D: number,
  q: Complex,
): Complex {
  const Aq: Complex = { re: A * q.re, im: A * q.im };
  const Bc: Complex = { re: B, im: 0 };
  const Cq: Complex = { re: C * q.re, im: C * q.im };
  const Dc: Complex = { re: D, im: 0 };
  const numer = cAdd(Aq, Bc);
  const denom = cAdd(Cq, Dc);
  return cDiv(numer, denom);
}

// ---------------------------------------------------------------------------
// Lens op
// ---------------------------------------------------------------------------

export const abcdThinLensOp: PhysicsOp = (
  rayIn: BeamRay,
  ctx: PhysicsOpContext,
): BeamRay[] => {
  // Prefer the transition's explicit 5×5 matrix when supplied — this is
  // how plano-convex lenses encode their direction-aware thick-lens
  // matrices (convex→flat vs flat→convex produce different BFL).
  // Falls back to the symmetric thin-lens matrix derived from focalLengthMm
  // when no explicit matrix is provided (simple use case).
  let M5: number[];
  if (ctx.transferMatrix?.kind === "matrix5x5") {
    M5 = ctx.transferMatrix.M;
  } else {
    const focalMm = (ctx.params.focalLengthMm ?? ctx.params.focalMm) as number | undefined;
    if (typeof focalMm !== "number" || !isFinite(focalMm) || focalMm === 0) {
      throw new Error(
        `abcd_thin_lens: needs transition.matrix5x5 OR valid focalLengthMm; got focal=${JSON.stringify(focalMm)}`,
      );
    }
    M5 = mThinLens(focalMm);
  }
  // 5×5 row-major: row r, col c → M5[r*5 + c]
  // x sub-block: (0,0)=A, (0,1)=B, (1,0)=C, (1,1)=D
  const Ax = M5[0 * 5 + 0], Bx = M5[0 * 5 + 1];
  const Cx = M5[1 * 5 + 0], Dx = M5[1 * 5 + 1];
  // y sub-block: (2,2)=A, (2,3)=B, (3,2)=C, (3,3)=D
  const Ay = M5[2 * 5 + 2], By = M5[2 * 5 + 3];
  const Cy = M5[3 * 5 + 2], Dy = M5[3 * 5 + 3];

  const qxOut = applyAbcdToQ(Ax, Bx, Cx, Dx, rayIn.qx);
  const qyOut = applyAbcdToQ(Ay, By, Cy, Dy, rayIn.qy);

  // Propagate chief ray from face_in to face_out (lens thickness in body-local).
  // For a thin lens, A and B faces sit at ±t/2; we propagate the ray's origin
  // along its current direction by the geometric distance between the two
  // face positions (approximate — thin-lens limit).
  const thicknessMm = vec3Distance(
    ctx.faceIn.positionMmBodyLocal,
    ctx.faceOut.positionMmBodyLocal,
  );
  const newOrigin = vec3Add(rayIn.origin, vec3Scale(rayIn.direction, thicknessMm));

  return [{
    ...rayIn,
    origin: newOrigin,
    qx: qxOut,
    qy: qyOut,
    pathLengthMm: rayIn.pathLengthMm + thicknessMm,
    // direction, jones, wavelength, power, phase: unchanged
  }];
};

// ---------------------------------------------------------------------------
// Registration
// ---------------------------------------------------------------------------

registerKind("lens", {
  ops: {
    abcd_thin_lens: abcdThinLensOp,
  },
  needsAperture: true,
});
